src/r_plane.c

```diff
@@ -376,152 +382,342 @@ static void R_MakeSpans(int x, unsigned int t1, unsigned int b1, unsigned int t2
-
-	    // Sky transferred from first sidedef
-	    const side_t *s = *l->sidenum + sides;
-
-	    if (s->baserowoffset - s->oldrowoffset)
-	      vertically_scrolling = true;
-
-	    // Texture comes from upper texture of reference sidedef
-	    texture = texturetranslation[s->toptexture];
-
-	    // Horizontal offset is turned into an angle offset,
-	    // to allow sky rotation as well as careful positioning.
-	    // However, the offset is scaled very small, so that it
-	    // allows a long-period of sky rotation.
-
-	    an += s->textureoffset;
-
-	    // Vertical offset allows careful sky positioning.
-
-	    dc_texturemid = s->rowoffset - 28*FRACUNIT;
-
-	    // We sometimes flip the picture horizontally.
-	    //
-	    // Doom always flipped the picture, so we make it optional,
-	    // to make it easier to use the new feature, while to still
-	    // allow old sky textures to be used.
-
-	    flip = l->special==272 ? 0u : ~0u;
-	  }
-	else 	 // Normal Doom sky, only one allowed per level
-	  {
-	    dc_texturemid = skytexturemid;    // Default y-offset
-	    texture = skytexture;             // Default texture
-	    flip = 0;                         // Doom flips it
-	  }
-
-        // Sky is always drawn full bright, i.e. colormaps[0] is used.
-        // Because of this hack, sky is not affected by INVUL inverse mapping.
-	//
-	// killough 7/19/98: fix hack to be more realistic:
-
-	if (STRICTMODE_COMP(comp_skymap) || !(dc_colormap[0] = dc_colormap[1] = fixedcolormap))
-	  dc_colormap[0] = dc_colormap[1] = fullcolormap;          // killough 3/20/98
-
-        dc_texheight = textureheight[texture]>>FRACBITS; // killough
-        dc_iscale = skyiscale;
-
-        // [FG] stretch short skies
-        stretch = (stretchsky && dc_texheight < 200);
-        if (stretch || !vertically_scrolling)
+    byte *rover = fire_pixels;
+    for (int x = 0; x < FIRE_WIDTH; x++)
+    {
+        byte *src = fire_indexes + x;
+        for (int y = 0; y < FIRE_HEIGHT; y++)
+        {
+            *rover++ = fire->palette[*src];
+            src += FIRE_WIDTH;
+        }
+    }
+}
+
+static void SpreadFire(void)
+{
+    for(int x = 0; x < FIRE_WIDTH; ++x)
+    {
+        for (int y = 1; y < FIRE_HEIGHT; ++y)
+        {
+            int src = y * FIRE_WIDTH + x;
+
+            int index = fire_indexes[src];
+
+            if (!index)
+            {
+                fire_indexes[src - FIRE_WIDTH] = 0;
+            }
+            else
+            {
+                int rand_index = M_Random() % 3;
+                int dst = src - rand_index + 1;
+                fire_indexes[dst - FIRE_WIDTH] = index - (rand_index & 1);
+            }
+        }
+    }
+}
+
+void R_SetupFire(fire_t *fire)
+{
+    memset(fire_indexes, 0, FIRE_WIDTH * FIRE_HEIGHT);
+
+    int last = array_size(fire->palette) - 1;
+
+    for (int i = 0; i < FIRE_WIDTH; ++i)
+    {
+        fire_indexes[(FIRE_HEIGHT - 1) * FIRE_WIDTH + i] = last;
+    }
+
+    for (int i = 0; i < 64; ++i)
```

Comment:
64 is FIRE_WIDTH?

Reply:
This is a requirement of the spec.
> Fire skies are expected to be primed with both the bottom row set to the last palette index; and with a minimum of 64 iterations performed to populate the target image with some fire to begin with.

https://docs.google.com/document/d/1RWNjk3mr3CLFptGCSVkZl5hFEPr1R5OzeSu7aHY8Zwg/edit?usp=sharing